hera_corr_cm/hera_corr_cm.py

```diff
@@ -786,6 +792,54 @@ def get_x_status(self):
         """Return a dictionary of X-engine status flags."""
         raise NotImplementedError("No code for get_x_status.")
 
+    def get_hashpipe_status(self):
+        """Return a dictionary of hashpipe status flags from redis.
+
+        Returns
+        -------
+        dict of lists
+            Upper lever dictionary is keyed by hashpipe keys in redis.
+            The value for each key is a list of dicts of all measurements in that redis hash.
+            Sub dictionaries are formatted to be properly ingested by an influxdb as json dicts.
+
+        """
+        rv = {}
+        for key in self.r.scan_iter("hashpipe:*/status"):
+            data_points = rv.setdefault(key, [])
+            _, _, host, pipeline, _ = key.split("/")
+            vals = self.r.hgetall(key)
+            timestamp = datetime.datetime.utcnow().isoformat()
```

Comment:
Yeah i remember this discussion. But when I print out what is listed in redis, there is no timestamp entry, or entry that looks like a time stamp. other than the sync time on the storage node, but I do not believe that would fit our need.
```
In [11]: keys[0], stats                                                                                                                                                                                             
Out[11]: 
(b'hashpipe://px8/1/status',
 {b'INSTANCE': b'1',
  b'BINDHOST': b'eth5',
  b'GPUDEV': b'1',
  b'XID': b'15',
  b'NETSTAT': b'running',
  b'GIT_VER': b'v2.0-29-g6da12078b0',
  b'BINDPORT': b'8511',
  b'MISSEDFE': b'0',
  b'MISSEDPK': b'0',
  b'FLUFSTAT': b'blocked_in',
  b'GPUSTAT': b'blocked_in',
  b'BDASTAT': b'blocked_in',
  b'OUTSTAT': b'blocked',
  b'OUTDUMPS': b'0',
  b'BDACONF': b'/tmp/bdaconfig.txt',
  b'INTSTAT': b'start',
  b'INTSYNC': b'6086656',
  b'INTCOUNT': b'36864',
  b'GPUDUMPS': b'0',
  b'GPUNANT': b'192',
  b'GPUNCHAN': b'384',
  b'GPUNTIME': b'2048',
  b'FLUFMCNT': b'0',
  b'TIMEIDX': b'0',
  b'NBL2SEC': b'0',
  b'NBL4SEC': b'0',
  b'NBL8SEC': b'4560',
  b'NBL16SEC': b'0',
  b'NBDABLS': b'9120',
  b'BDASAMP': b'0',
  b'NETWATMN': b'99999',
  b'NETWATMX': b'0',
  b'NETRECMN': b'99999',
  b'NETRECMX': b'0',
  b'NETPRCMN': b'99999',
  b'NETPRCMX': b'0'})
```
```
In [15]: key, r.hgetall(key)                                                                                                                                                                                        
Out[15]: 
(b'hashpipe://hera-sn1/0/status',
 {b'INSTANCE': b'0',
  b'BINDHOST': b'eth4',
  b'CNETSTAT': b'running',
  b'GIT_VER': b'v2.0-29-g6da12078b0',
  b'BINDPORT': b'10000',
  b'MISSXENG': b'0',
  b'MISSEDPK': b'0',
  b'DISKSTAT': b'blocked_in',
  b'AUTOSTAT': b'init',
  b'DISKMCNT': b'0',
  b'TRIGGER': b'1',
  b'NDONEFIL': b'0',
  b'NFILES': b'186',
  b'INTTIME': b'36864',
  b'SYNCTIME': b'1591651632000',
  b'HDF5TPLT': b'/tmp/template.h5',
  b'NETWATMN': b'99999',
  b'NETWATMX': b'0',
  b'NETRECMN': b'99999',
  b'NETRECMX': b'0',
  b'NETPRCMN': b'99999',
  b'NETPRCMX': b'0',
  b'NBL2SEC': b'0',
  b'NBL4SEC': b'0',
  b'NBL8SEC': b'4560',
  b'NBL16SEC': b'0',
  b'BDANANT': b'95',
  b'MSPERFIL': b'19327',
  b'TAG': b'engineering'})
```

Reply:
I found a lot more entries when I did this awhile back (not necessarily time tags) but I'm curious why.  For the px one it could be that I aggregated all of the px<1-16>/<0,1> entries but that's not the case for sn1.

Reply:
Could be the last time this was updated it didn't fill out everything from your other list? Not sure where this gets populated myself.

Reply:
Comparing with the list you sent me, nothing looks like a timestamp or time tag :thinking:

Reply:
@david-macmahon I haven't been able to find anything that looks like a `timestamp` in redis under any of the hashpipe stats. Do you know if this particular field is getting ported up all the way?

Reply:
Looks like `SYNCTIME` (in milliseconds) from the catcher machine (`hera-sn1`) should probably be added to the status buffers on all hosts.  It should be easy enough to publish that value "globally" rather than just to the catcher(s?).  It wasn't clear which (if any) of these values represents the `MCOUNT` (aka sample/spectrum/packet counter) of the block most recently passed from the net thread to the next thread.  That should be relatively easy to add (if that's not in there already and I'm just not seeing which field it is).

From those two values, the real UTC time can be derived from:
```
 UTC_MS = SYNCTIME + MCOUNT / MCOUNT_PER_MS
```
where `MCOUNT_PER_MS` is a rational constant whose value depends on the FPGA's implementation of `MCOUNT` (i.e. Jonathon can provide that value).  FWIW, the value of `SYNCTIME` shown above corresponds to 2020-06-08T21:27:12.  That is essentially T=0 for the correlator.

Reply:
Depending on the value, you might want to store `MCPERMS` (or `MSPERMC` if you prefer to multiply) as a stringified fraction to the status buffers (or use `NCPERMSN` and `MSOERMSD` for the numerator/denominator of that rational number).

Reply:
I don't think the choice of multiply or divide really makes a big difference for me personally, just need to know which we're going to use. I think my mind likes `MCPERMS` better. We could just use the calculation you mentioned, and the relevant information from the catcher node hashpipes to create a global timestamp for all data anyway.

For the record this is how influxdb handles duplicate point:
>A point is uniquely identified by the measurement name, tag set, and timestamp. If you submit a new point with the same measurement, tag set, and timestamp as an existing point, the field set becomes the union of the old field set and the new field set, where any ties go to the new field set. This is the intended behavior.
so it looks like duplicates (if exactly duplicate) will be ignored.

@david-macmahon is there a possible scenario where the `MCOUNT` on the catcher node has not changed, but the actual statistics could be different? this would cause us to overwrite data from the old `MCOUNT`. Naively I would guess this scenario would not be very applicable at the second or 5 second time scales.

Reply:
When hashpipe is processing data there are more redis keys. Here is the set of redis keys for the px machines and hera-sn1 (catcher).

```
In [11]: r.hgetall('hashpipe://px12/1/status')                                        
Out[11]: 
{b'INSTANCE': b'1',
 b'BINDHOST': b'eth5',
 b'GPUDEV': b'1',
 b'XID': b'23',
 b'NETSTAT': b'running',
 b'GIT_VER': b'v2.0-29-g6da12078b0',
 b'BINDPORT': b'8511',
 b'MISSEDFE': b'180',
 b'MISSEDPK': b'0',
 b'FLUFSTAT': b'fluffing',
 b'GPUSTAT': b'waiting',
 b'BDASTAT': b'blocked_in',
 b'OUTSTAT': b'blocked',
 b'OUTDUMPS': b'2',
 b'BDACONF': b'/tmp/bdaconfig.txt',
 b'INTSTAT': b'on',
 b'INTSYNC': b'25419778',
 b'INTCOUNT': b'32768',
 b'GPUDUMPS': b'22',
 b'GPUNANT': b'192',
 b'GPUNCHAN': b'384',
 b'GPUNTIME': b'2048',
 b'FLUFMCNT': b'26906626',
 b'TIMEIDX': b'1',
 b'NBL2SEC': b'0',
 b'NBL4SEC': b'0',
 b'NBL8SEC': b'78',
 b'NBL16SEC': b'0',
 b'NBDABLS': b'156',
 b'BDASAMP': b'21',
 b'NETBKOUT': b'9',
 b'FLUFBKIN': b'9',
 b'NETMCNT': b'26906626',
 b'NETGBPS': b'41.790317535',
 b'NETWATNS': b'31653.765625000',
 b'NETRECNS': b'78.572265625',
 b'NETPRCNS': b'805.077392578',
 b'NETWATMN': b'13',
 b'NETRECMN': b'25',
 b'NETPRCMN': b'22',
 b'NETWATMX': b'6860991',
 b'NETRECMX': b'1748973',
 b'NETPRCMX': b'1936887',
 b'NETPKTS': b'4096',
 b'NETDROPS': b'0',
 b'NETPKTTL': b'6494996',
 b'NETDRPTL': b'135832',
 b'FLUFGBPS': b'58.049587250',
 b'FLUFMING': b'13.867094994',
 b'GPUBLKIN': b'0',
 b'GPUMCNT': b'26902530',
 b'GPUGBPS': b'93.014984131',
 b'BDABLKIN': b'1',
 b'BDAMCNT': b'26828802',
 b'BDABLKOU': b'0',
 b'BDASECS': b'0.001096951',
 b'BDAMBPS': b'1747.505615234',
 b'OUTBLKIN': b'1',
 b'OUTBCNT': b'134217848',
 b'OUTBYTES': b'1928160',
 b'OUTSECS': b'0.043935243',
 b'OUTMBPS': b'351.091247559'}
```


```
In [19]: r.hgetall('hashpipe://hera-sn1/0/status')                                    
Out[19]: 
{b'INSTANCE': b'0',
 b'BINDHOST': b'eth4',
 b'CNETSTAT': b'running',
 b'GIT_VER': b'v2.0-29-g6da12078b0',
 b'BINDPORT': b'10000',
 b'MISSXENG': b'0',
 b'MISSEDPK': b'0',
 b'DISKSTAT': b'blocked_in',
 b'AUTOSTAT': b'writing',
 b'DISKMCNT': b'29843456',
 b'TRIGGER': b'0',
 b'NDONEFIL': b'4',
 b'HDF5TPLT': b'/tmp/template.h5',
 b'NFILES': b'10',
 b'SYNCTIME': b'1592412682000',
 b'INTTIME': b'32768',
 b'NETWATMN': b'32',
 b'NETWATMX': b'36675315056',
 b'NETRECMN': b'37',
 b'NETRECMX': b'30628',
 b'NETPRCMN': b'1189',
 b'NETPRCMX': b'279798',
 b'NBL2SEC': b'0',
 b'NBL4SEC': b'0',
 b'NBL8SEC': b'78',
 b'NBL16SEC': b'0',
 b'BDANANT': b'12',
 b'DISKBKIN': b'1',
 b'DISKBCNT': b'1280',
 b'NETBKOUT': b'1',
 b'NETBCNT': b'1280',
 b'NETGBPS': b'15.280817032',
 b'NETWATNS': b'1395587.125000000',
 b'NETRECNS': b'169.998458862',
 b'NETPRCNS': b'2246.626220703',
 b'NETPKTS': b'24567',
 b'NETDROPS': b'0',
 b'NETPKTTL': b'171939',
 b'NETDRPTL': b'0',
 b'AUTOBKIN': b'1',
 b'MSPERFIL': b'60000',
 b'TAG': b'test',
 b'FILESEC': b'17.179868698',
 b'DISKTBNS': b'320546.875000000',
 b'DISKTMIN': b'5118129',
 b'DISKTMAX': b'25444673',
 b'DISKWBNS': b'415704.875000000',
 b'DISKWMIN': b'335392',
 b'DISKWMAX': b'25378274',
 b'DISKWBL': b'38507',
 b'DISKGBPS': b'0.752599478',
 b'DUMPMS': b'267.508270264',
 b'DISKMING': b'0.524775088'}
```

Of those the variables that are important fall into two categories- those that are constant in time and those that change with time so have to be polled more often.

### pxN machines
Constant:
GPUNANT, GPUNCHAN, GPUNTIME, 
NBL2SEC, NBL4SEC, NBL8SEC, NBL16SEC, 
NBDABLS

Time Variable:
MISSEDFE, 
MISSEDPK, 
NETDRPTL, NETDROPS
NETGBPS, FLUFGBPS, GPUGBPS, BDAMBPS, OUTMBPS

### hera-sn1 (catcher)
Constant:
BDANANT. FILESEC

Time Variable:
MISSXENG, 
MISSEDPK,
NETDRPTL, NETDROPS
NDONEFIL, NFILES
AUTOBKIN
NETGBPS, DISKGBPS
DISKTMIN, DISKTMAX, DISKWMIN, DISKWMAX

The variables on the same line can tentatively be on the same panel because either they y-axis has the same range or because they logically make sense.

Reply:
To obtain the timestamp in unix time:

```
sync_time_ms = r.hget('hashpipe://hera-sn1/0/status', 'SYNCTIME')
mcnt = r.hget('hashpipe://hera-sn1/0/status', 'DISKMCNT')
unix_time = (sync_time_ms / 1000.) + (mcnt * 8192 / 500e6)
```

Here 8192 is the total number of channels generated and 500e6 is the sampling frequency of the ADCs.

Reply:
That's great. Will `DISKMCNT` retain the value of the last time stamp or do we need a special way to handle when the hashpipes are no longer taking data?

Reply:
It will retain the last value if the hashpipe instance hangs or is not killed. I think it gets deleted when the hashpipe instance is shutdown.

Reply:
@mkolopanis That's a good point about `MCOUNT` maybe not changing.  We wouldn't want a non-advancing MCOUNT to stop time.  I think I'd like to retract my recommendation to use `MCOUNT` and `SYNCTIME` for the influxDB timestamps. The timestamps for that purpose should either come from the `hashpipe_redis_gateway` process (timestamping the time the data are written to Redis) or the redis-to-influxDB process (timestamping the time the data were pulled from Redis).  Not sure if one would be preferable to the other.  Like a stopped `MCOUNT`, we also wouldn't want a stopped `hashpipe_redis_gateway` to stop time so maybe adding the timestamp at the redis-to-influxDB time would be most useful.

Reply:
(I still think the synctime-mcount-to-unix-time function is a good idea, even if not for this particular purpose)

Reply:
@bhazelton looks like one of dave's new proposed methods for timestamps is just what this code already does (making a timestamp at execution time). Could look to change it to the other way (getting from `hashpip_redis_gateway` though), but then i also still worry about that one stopping or such.

Reply:
Ok. I'm actually not worried about "time stopping" in the sense that if we stop getting new data points when the data aren't changing that's appropriate and lets us know there's a problem. Getting a lot of copies of the same points with new timestamps is in some way more confusing and that's what I'm worried about with this implementation.

Reply:
@dgorthi commented in slack today about using the `auto:timestamp` field in redis, which is populated itself by the hashpipes. This could be a viable solution as well.

Reply:
That seems reasonable as long as it is appropriate to associate the data in the hashpipe keys with that timestamp (i.e. they are populated at roughly the same time and it can't happen that either the timestamp or data changes when the other does not)

Reply:
In my mind, the most natural solution is to have whatever puts data into the hashpipe keys in redis write a timestamp that is the time that data was copied from the hashpipe.

Reply:
@bhazelton Getting continuously timestamped data points when values are not changing tells you that the process updating influxDB is still working and the unchanging values mean that something upstream is not behaving as expected (i.e. has "flat lined").  Getting no new values if they are not changing is ambiguous.  No updates in that case could mean either the process updating influxDB is not working or the upstream processes are not behaving as expected (or both!)

Reply:
I think there are better ways to make sure the process updating the influxDB is working (e.g. a heartbeat). Old data with new timestamps is just confusing.

Reply:
Good point, I agree about the heartbeat being a better way to ensure that the ingestion process is running.  I guess the question about the merits of unchanged values with new timestamps depends on how the database differentiates between non-changing and missing values.  If a value was Y at time T0 and doesn't change until it disappears at time T1, how would that disappearance be reflected in the database?  In the only-record-new-timestamp-on-value-changed method, the DB would need to record a transition to NULL at time T1.  In the record-new-timestamps-periodically-for-data-points-that-have-any-value method, the data point would be present (albeit unchanging) for all timestamps until it ceases to exist at time T1.  I'm not familiar with how influxDB handles missing data, so I don't know which way would be best.

Reply:
I think with how i've set it up, this _should_ show up on the grafana M&C script monitoring dashboard. The `auto:timestamp` gets created by the hashpipes, so presumably it is relevant for the data being read out of the status :thinking: @dgorthi  would you think this is reasonable time association?

Reply:
I actually agree with @david-macmahon. We don't want a stuck mcount to stop time (like @mkolopanis pointed out), and this same problem exists with auto:timestamp as well. I should've read all the comments more carefully. I agree that we should have a record-new-timestamp-periodically-for-datapoints-that-have-any-value method, which is the method currently implemented.

@bhazelton The processes that put the variables in redis are individual hashpipe threads. Each thread uploads/changes some keys every time a particular process-loop is executed. This time cadence for this is often a few milliseconds. Even if the threads uploaded a timestamp, this timestamp not changing would be useful information because it would mean the pipeline hanged. From a debugging standpoint it is useful to upload keys that are not changing as well- including mcount and synctime which timestamp data and other processes.

Moreover, since the influx database is finite (carry a week or two worth of data) populating it periodically, even if with the same values, shouldn't cause harm.

Reply:
Just to be explicitly clear, the current retention policy is 4 weeks, but can easily be reduced if desired.

Reply:
I think the solution in this PR is the right one given that different threads are writing the various values at different times. We just need to remember that the timestamps in this monitor are the times they are read from redis, not the times they are written to redis. Maybe we can make that clear in the influx interface.